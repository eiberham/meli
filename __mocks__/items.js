export const items = [{
    "id": "MLA619208970",
    "title": "Remera Lisa Algodón Jersey Peinado Premium",
    "condition": "new",
    "picture": "http://mla-s1-p.mlstatic.com/876978-MLA31035095700_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 259,
        "decimals": ""
    }
}, {
    "id": "MLA737922842",
    "title": "Remeras Personalizadas Estampada Sublimada Tu Logo",
    "condition": "new",
    "picture": "http://mla-s1-p.mlstatic.com/974505-MLA31102958584_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 199.99,
        "decimals": "99"
    }
}, {
    "id": "MLA752716977",
    "title": "Remeras De Algodon Premium Al Por Mayor",
    "condition": "new",
    "picture": "http://mla-s2-p.mlstatic.com/828346-MLA31115819250_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 260,
        "decimals": ""
    }
}, {
    "id": "MLA670757536",
    "title": "Remeras Entalladas Hombre De Importación Slim Fit!!",
    "condition": "new",
    "picture": "http://mla-s1-p.mlstatic.com/708594-MLA31133445654_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 249.99,
        "decimals": "99"
    }
}];