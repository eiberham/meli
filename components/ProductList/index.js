import React, { setState } from 'react';
import './product-list.scss';
import PropTypes from 'prop-types';

import ProductItem from '../ProductItem';

const ProductList = props => {
    const { items, onItemSelect } = props;

    const onProductSelect = ({id}) => {
        onItemSelect(id);
    }

    return (
        <React.Fragment>
            {items && items.map((item, index) => (
                <ProductItem 
                    key={index} 
                    product={item} 
                    onProductSelect={() => onProductSelect(item)} 
                />
            ))}content
        </React.Fragment>
    );
};

ProductList.propTypes = {
    items: PropTypes.arrayOf(
                PropTypes.shape({
                    id: PropTypes.string,
                    title: PropTypes.string,
                    condition: PropTypes.string,
                    free_shipping: PropTypes.bool,
                    picture: PropTypes.string,
                    price: PropTypes.shape({
                        amount: PropTypes.number,
                        currency: PropTypes.string,
                        decimals: PropTypes.string
                    })

                }).isRequired
            ).isRequired
}

export default ProductList;