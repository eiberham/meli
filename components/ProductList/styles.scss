.products {
    background: #fff;
    min-width: 100%;
    margin: 0 auto;
    box-shadow: 0 0 5px rgba(33,33,33,.2);
}