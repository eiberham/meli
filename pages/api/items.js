export default (req, res) => {
    console.log("el objeto req negro: ", req);
    res.send("Todo bien papu");
}