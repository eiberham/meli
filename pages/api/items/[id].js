import redis from 'redis';
import axios from 'axios';
import _ from 'lodash';
import meli from '../../../services/meli';

export default (req, res) => {
    const client = redis.createClient();
    let id = req.query.id;

    client.get(id, async (err, result) => {
        if (result) {
            res.send(result);
        } else {
            axios.all([
                meli.get(`/items/${id}`),
                meli.get(`/items/${id}/description`)
            ])
                .then(axios.spread(async (first, second) => {
                    let response = {...first.data, ...second.data};

                    let data = {
                        ..._.pick(response, ['id', 'title', 'condition', 'sold_quantity']),
                        picture: _.get(response, 'thumbnail'),
                        free_shipping: _.get(response, 'shipping.free_shipping'),
                        description: _.get(response, 'plain_text'),
                        price: {
                            currency: _.get(response, 'currency_id'),
                            amount: _.get(response, 'price'),
                            decimals: _.toString(_.split(_.get(response, 'price'), '.')[1])
                        }
                    };

                    const {data:{path_from_root}} = await meli.get(`/categories/${response.category_id}`);
                    let categories = _.map(path_from_root, item => {
                        return item.name
                    });

                    let output = {
                        author: { name: 'Abraham', lastname: 'Cedeno'},
                        item: data,
                        categories
                    };

                    client.setex(id, 3600, JSON.stringify(output));

                    res.status(200).json(output);
                }))
                .catch(({response:{data}}) => {
                    if(data.status === 404){
                        return res.status(404).json({
                            "status": 404,
                            "message": data.message
                        });
                    }
                });
        }
    });
}