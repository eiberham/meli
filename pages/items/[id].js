import fetch from 'isomorphic-unfetch';
import Layout from '../../components/Layout';
import Breadcrumb from '../../components/Breadcrumb';
import ProductDetail from '../../components/ProductDetail';

const Item = props => {
    const { item, categories } = props;

    return (
        <Layout>
            <Breadcrumb links={categories} />
            <ProductDetail item={item} />
        </Layout>
    )
}

Item.getInitialProps = async context => {
    const {
        query: { id },
    } = context;
    
    const response = await fetch(`http://localhost:3000/api/items/${id}`);
    const data = await response.json();
    return data;
}

export default Item;