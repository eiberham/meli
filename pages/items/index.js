import { Fragment } from 'react';
import fetch from 'isomorphic-unfetch';
import Layout from '../../components/Layout';
import Breadcrumb from '../../components/Breadcrumb';
import ProductList from '../../components/ProductList';
import Loading from '../../components/Loading';
import { withRouter } from 'next/router';

const Items = props => {
    const { items, router, categories } = props;

    const onItemSelect = (id) => {
        router.push(`/items/${id}`)
    }

    return (
        <Layout>
            { (items) ? (
                <Fragment>
                    <Breadcrumb links={categories} />
                    <ProductList items={items} onItemSelect={onItemSelect}/>
                </Fragment>
            ) : (
                <Loading size="massive" show />
            )}
        </Layout>
    );
};

Items.getInitialProps = async (context) => {
    const {
        query: { search },
    } = context;
    
    const response = await fetch(`http://localhost:3000/api/items?q=${search}`);
    const data = await response.json();
    return data;
}

export default withRouter(Items);