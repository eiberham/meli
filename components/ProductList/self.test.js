import React from 'react';
import { items } from '../../__mocks__/items';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ProductList from './index';

describe('ProductList', () => {
    it('matches snapshot', () => {
        const onItemSelect = jest.fn();
        const component = renderer.create(<ProductList items={items} onItemSelect={onItemSelect} />)
        .toJSON();
        expect(component).toMatchSnapshot();
    })

    it('works like a charm', () => {
        const onItemSelect = jest.fn();
        const component = mount(<ProductList items={items} onItemSelect={onItemSelect} />);
        expect(component.find('.item')).toHaveLength(4);
        component.find('.item').at(0).simulate('click');
        expect(onItemSelect).toHaveBeenCalledTimes(1);
    })
})