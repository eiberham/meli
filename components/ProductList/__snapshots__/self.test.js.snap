// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList matches snapshot 1`] = `
<div
  className="products"
>
  <div
    className="item"
    onClick={[Function]}
  >
    <div
      className="item-image"
    >
      <img
        alt="Remera Lisa Algod贸n Jersey Peinado Premium"
        src="http://mla-s1-p.mlstatic.com/876978-MLA31035095700_062019-I.jpg"
      />
    </div>
    <div
      className="item-description"
    >
      <div
        className="item-price"
      >
        $ 
        259
        <span
          className="item-decimals"
        >
          
        </span>
      </div>
      <div
        className="item-title"
      >
        Remera Lisa Algod贸n Jersey Peinado Premium
      </div>
    </div>
    <div
      className="item-condition"
    >
      <span>
        Nuevo
      </span>
    </div>
  </div>
  <div
    className="item"
    onClick={[Function]}
  >
    <div
      className="item-image"
    >
      <img
        alt="Remeras Personalizadas Estampada Sublimada Tu Logo"
        src="http://mla-s1-p.mlstatic.com/974505-MLA31102958584_062019-I.jpg"
      />
    </div>
    <div
      className="item-description"
    >
      <div
        className="item-price"
      >
        $ 
        199
        <span
          className="item-decimals"
        >
          99
        </span>
      </div>
      <div
        className="item-title"
      >
        Remeras Personalizadas Estampada Sublimada Tu Logo
      </div>
    </div>
    <div
      className="item-condition"
    >
      <span>
        Nuevo
      </span>
    </div>
  </div>
  <div
    className="item"
    onClick={[Function]}
  >
    <div
      className="item-image"
    >
      <img
        alt="Remeras De Algodon Premium Al Por Mayor"
        src="http://mla-s2-p.mlstatic.com/828346-MLA31115819250_062019-I.jpg"
      />
    </div>
    <div
      className="item-description"
    >
      <div
        className="item-price"
      >
        $ 
        260
        <span
          className="item-decimals"
        >
          
        </span>
      </div>
      <div
        className="item-title"
      >
        Remeras De Algodon Premium Al Por Mayor
      </div>
    </div>
    <div
      className="item-condition"
    >
      <span>
        Nuevo
      </span>
    </div>
  </div>
  <div
    className="item"
    onClick={[Function]}
  >
    <div
      className="item-image"
    >
      <img
        alt="Remeras Entalladas Hombre De Importaci贸n Slim Fit!!"
        src="http://mla-s1-p.mlstatic.com/708594-MLA31133445654_062019-I.jpg"
      />
    </div>
    <div
      className="item-description"
    >
      <div
        className="item-price"
      >
        $ 
        249
        <span
          className="item-decimals"
        >
          99
        </span>
      </div>
      <div
        className="item-title"
      >
        Remeras Entalladas Hombre De Importaci贸n Slim Fit!!
      </div>
    </div>
    <div
      className="item-condition"
    >
      <span>
        Nuevo
      </span>
    </div>
  </div>
</div>
`;
